Parallel matrix multiplication

bulat-tsydendorzhiev opened the pull request.

bulat-tsydendorzhiev committed:
feat: add matrix class, matrix multiplier with sequential matrix multiplication, simple tests

bulat-tsydendorzhiev committed:
docs: add documentation for writting method

bulat-tsydendorzhiev committed:
refactor: merge sequential and parallel multipliers into one class, replace List matrix to array

bulat-tsydendorzhiev committed:
feat: add parallel multiplication; fix: fixed equation method; refactor: change TestDataClass data for valid cases

bulat-tsydendorzhiev committed:
feat: add output file parameter for user; fix: fixed parallel multiplier

bulat-tsydendorzhiev committed:
feat: add Benchmark class

bulat-tsydendorzhiev committed:
feat: add StyleCop to project

bulat-tsydendorzhiev committed:
feat: change creating a table with results

bulat-tsydendorzhiev committed:
feat: add StyleCop

bulat-tsydendorzhiev committed:
style: make style by StyleCop preferences

bulat-tsydendorzhiev committed:
Add missed file

yurii-litvinov commented on ParallelMatrixMultiplication/Benchmark/Benchmark.cs:
```diff
@@ -0,0 +1,147 @@
+// <copyright file="Program.cs" company="bulat-tsydendorzhiev">
```
Тут StyleCop ругается, что в комментарии неправда относительно имени файла

yurii-litvinov commented on ParallelMatrixMultiplication/Benchmark/Benchmark.cs:
```diff
@@ -0,0 +1,147 @@
+// <copyright file="Program.cs" company="bulat-tsydendorzhiev">
+// Copyright (c) bulat-tsydendorzhiev. All Rights Reserved.
+// Licensed under the MIT License. See LICENSE in the repository root for license information.
+// </copyright>
```
Надо неймспейс (в .NET всё, кроме Main, должно быть в каком-нибудь неймспейсе, при этом корневой неймспейс должен совпадать по имени с именем проекта).

yurii-litvinov replied:
Неймспейса нету

yurii-litvinov commented on ParallelMatrixMultiplication/Benchmark/Benchmark.cs:
```diff
@@ -0,0 +1,147 @@
+        if (matrixMultiplierName == "Parallel")
+        {
+            cells.Add(new PdfPCell(new Phrase(this.parallelResults[numberOfRow].ExpectedValue.ToString(), font)));
+            cells.Add(new PdfPCell(new Phrase(this.parallelResults[numberOfRow].StandartDeviation.ToString(), font)));
```
Тысячи цифр после запятой в представлении результата могут очень сильно расстроить Кирилла Константиновича, который по идее должен был рассказывать про постановку эксперимента. И то, что в матожидании то есть цифра после запятой, то нет (лучше 30.0, чем просто 30, если в остальных результатах есть дробная часть — это говорит, что десятые тоже измерены и равны нулю, а не попали в погрешность).

yurii-litvinov commented on ParallelMatrixMultiplication/ParallelMatrixMultiplication/Exceptions/DimensionsMismatchException.cs:
```diff
@@ -0,0 +1,27 @@
+    /// Initializes a new instance of the <see cref="DimensionsMismatchException"> exception.
+    /// </summary>
+    /// <param name="message">Throwing message.</param>
+    public DimensionsMismatchException(string message) : base(message)
```
Тут и везде стоит либо поправить предупреждения StyleCop, либо (что хуже) заглушить их. Собираться должно вообще без предупреждений.

yurii-litvinov commented on ParallelMatrixMultiplication/ParallelMatrixMultiplication/Matrix.cs:
```diff
@@ -0,0 +1,170 @@
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Matrix"/> class.
+    /// </summary>
+    /// <param name="newMatrix">Matrix with the ractangle form.</param>
```
```suggestion
    /// <param name="newMatrix">Matrix with the rectangle form.</param>
```
И из описания вообще не очевидно, что получившийся объект Matrix — это на самом деле view на массив (то есть если массив меняется, то и Matrix меняется). Я бы как пользователь ожидал, что матрица копируется внутрь. Стоило указать, что этого не происходит.

yurii-litvinov commented on ParallelMatrixMultiplication/ParallelMatrixMultiplication/Matrix.cs:
```diff
@@ -0,0 +1,170 @@
+    /// <param name="outputPath">Path where matrix will be located.</param>
+    public void WriteToFile(string outputPath)
+    {
+        using (var writer = new StreamWriter(outputPath, false, Encoding.UTF8))
```
```suggestion
        using var writer = new StreamWriter(outputPath, false, Encoding.UTF8);
```
И дальше без лишних отступов.

yurii-litvinov commented on ParallelMatrixMultiplication/ParallelMatrixMultiplication/MatrixMultiplier.cs:
```diff
@@ -0,0 +1,99 @@
+        }
+
+        var result = new int[firstMatrix.RowsNumber, secondMatrix.ColumnsNumber];
+        var threads = new Thread[Environment.ProcessorCount];
```
В матрице может быть меньше строк, чем на машине ядер, тогда лишние потоки создадутся. Тут можно поточнее.

bulat-tsydendorzhiev committed:
style: make style by StyleCop preferences

refactor: make copying of matrix in constructor overloading

fix: change calculation of expected value and standart deviation, fixed problem of extra threads

docs: fixed typos and corrected mistakes in desriptions

test: add test for invalid matrix for second constructor overloading

yurii-litvinov commented on ParallelMatrixMultiplication/.editorconfig:
```diff
@@ -0,0 +1,10 @@
+dotnet_diagnostic.SA1309.severity = none
+
+# SA1200: Using directives should be placed correctly
+dotnet_diagnostic.SA1200.severity = none
```
Почему нет, дело говорит же

yurii-litvinov commented on ParallelMatrixMultiplication/.editorconfig:
```diff
@@ -0,0 +1,10 @@
+dotnet_diagnostic.SA1101.severity = none
+
+# SA1309: Field '...' should not begin with an underscore
+dotnet_diagnostic.SA1309.severity = none
```
Да и это вполне хорошее правило

yurii-litvinov commented on ParallelMatrixMultiplication/.editorconfig:
```diff
@@ -0,0 +1,10 @@
+[*.cs]
+
+# SA1101: Prefix local calls with this
+dotnet_diagnostic.SA1101.severity = none
```
И даже это в целом ок, можно привыкнуть :)

yurii-litvinov reviewed (approved):
В целом всё ок, зачтена
